src/alvin0319/OffHand/OffHand.php

```diff
@@ -30,84 +30,88 @@
 use pocketmine\event\Listener;
 use pocketmine\event\player\PlayerDeathEvent;
 use pocketmine\event\player\PlayerJoinEvent;
-use pocketmine\event\player\PlayerQuitEvent;
 use pocketmine\event\server\DataPacketReceiveEvent;
 use pocketmine\event\server\DataPacketSendEvent;
+use pocketmine\item\enchantment\Enchantment;
 use pocketmine\item\Item;
-use pocketmine\nbt\tag\CompoundTag;
+use pocketmine\nbt\tag\ListTag;
 use pocketmine\network\mcpe\protocol\AddPlayerPacket;
 use pocketmine\network\mcpe\protocol\MobEquipmentPacket;
 use pocketmine\network\mcpe\protocol\types\ContainerIds;
+use pocketmine\entity\Human;
 use pocketmine\Player;
 use pocketmine\plugin\PluginBase;
 
 use function array_merge;
+use function array_filter;
 
 class OffHand extends PluginBase implements Listener{
 
-	/** @var OffHandInventory[] */
-	protected $inventories = [];
-
 	public function onEnable() : void{
 		$this->getServer()->getPluginManager()->registerEvents($this, $this);
 	}
-
+	/**
+	 * @priority MONITOR
+	 */
 	public function onPlayerJoin(PlayerJoinEvent $event) : void{
-		$this->inventories[$event->getPlayer()->getName()] = new OffHandInventory($event->getPlayer());
-		$this->loadInventory($event->getPlayer());
-	}
-
-	public function onPlayerQuit(PlayerQuitEvent $event) : void{
-		$this->saveInventory($event->getPlayer());
-	}
-
-	private function loadInventory(Player $player) : void{
-		$player->addWindow($this->getOffHandInventory($player), ContainerIds::OFFHAND, true);
-		if($player->namedtag->hasTag("OffHand", CompoundTag::class)){
-			$this->getOffHandInventory($player)->setItemInOffHand(Item::nbtDeserialize($player->namedtag->getCompoundTag("OffHand")));
-		}
-	}
-
-	private function saveInventory(Player $player) : void{
-		$player->namedtag->setTag($this->getOffHandInventory($player)->getItemInOffHand()->nbtSerialize(-1, "OffHand"));
-		unset($this->inventories[$player->getName()]);
+        	$player = $event->getPlayer();
+        	$this->getOffhandInventory($player)->sendContents($player);
 	}
 
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketReceive(DataPacketReceiveEvent $event) : void{
 		$packet = $event->getPacket();
 		$player = $event->getPlayer();
-		if($packet instanceof MobEquipmentPacket){
-			if($packet->windowId === ContainerIds::OFFHAND){
-				$inv = $this->getOffHandInventory($player);
-				if($inv instanceof OffHandInventory){
-					$inv->setItemInOffHand($packet->item);
-					$event->setCancelled();
-				}
-			}
-		}
+        	if($packet instanceof MobEquipmentPacket and $packet->windowId == ContainerIds::OFFHAND){
+            		$offhand = $this->getOffhandInventory($player);
+            		if(!$offhand->getItem(0)->equalsExact($packet->item)){
+                		$offhand->sendContents($player);
+                		$event->setCancelled();
+                		return;
+            		}
+            		$offhand->setItem(0, $packet->item);
```

Comment:
Add $event->setCancelled() here. PMMP doesn't handle except for offhand.

Reply:
I used that and it worked perfectly somehow

Reply:
The Player handleMobEquipment uses this too and it seems that works perfectly as pmmp would handle it in the Item In Hand

Reply:
No, see https://github.com/pmmp/PocketMine-MP/blob/stable/src/pocketmine/Player.php#L2791

This will set player hotbar even player equipped item on offhand